Adding "Image Align with Rife" and "Wavelet Color Fix" Nodes

pifroggi opened the pull request:
Hey! I was asked if I could contribute my "Image Align with Rife" node to the official chaiNNer repository. For this purpose I've created this pull request. In addition it also adds a "Wavelet Color Fix" node. The code was tested with Alpha v0.22.2 on Windows.
I've originally not contributed as I'm unfamiliar with the process (this is my first time using github) and I have only very limited coding abilities, for which I would like to apologize already in case I've done something wrong.

### Image Align with Rife
Aligns an Image with a Reference Image using a modified [Rife](https://github.com/megvii-research/ECCV2022-RIFE). Images should have vague alignment before using this Node. Output Image will have the same dimensions as Reference Image. Resize Reference Image to get desired output scale.

Examples (1.Input 2.Reference 3.Output): https://slow.pics/c/rqeq3D97

### Wavelet Color Fix
Correct for upscaling model color shift by first separating the image into wavelets of different frequencies, then matching the average color of the Input Image to that of a Reference Image. In general produces better results than the Average Color Fix at the cost of more computation. The Wavelet Color Fix functions are from [sd-webui-stablesr](https://github.com/pkuliyi2015/sd-webui-stablesr/blob/master/srmodule/colorfix.py).

Example transferring DVD colors to upscaled image: https://imgsli.com/MjM5NzM5/0/2

### Example Chain
An example chain that produces good alignment in general can be found here:
https://github.com/pifroggi/custom_chainner_nodes/tree/main/Examples

I hope everything makes sense!

pifroggi commented:
Yes, I can try that! The official download would be a zip file hosted on google drive that contains additional files as well.
https://drive.google.com/file/d/1BjuEY7CHZv1wzmwXSQP9ZTj0mLWu_4xy/view
Should we use that, or host it here on github maybe?

joeyballentine commented:
The Google drive link should work, you'll just have to convert it to a direct download link first

pifroggi commented:
The model now gets automatically downloaded into chainner's "python/models/rife_v4.14" if it's not already there. Let me know if this path works.

joeyballentine commented:
Sorry for the delay, will review this when i have some time

pifroggi commented:
No worries, thanks! :)

joeyballentine commented:
The code for this looks fine to me. I'll test it out today and just verify everything works. thanks again for this

pifroggi commented:
Great, thank you as well!

joeyballentine commented on backend/src/packages/chaiNNer_pytorch/pytorch/processing/image_align_rife.py:
```diff
@@ -0,0 +1,211 @@
+
+@processing_group.register(
+    schema_id="chainner:pytorch:image_align_rife",
+    name="Image Align with Rife",
```
do you mind renaming this to "Align Image to Reference"

pifroggi replied:
No problem. My thinking with mentioning Rife in the name was in case in the future different kinds of alignment methods are added. For example it may make sense to add a traditional affine alignment method at some point, which would be better with images that are off by a lot, which could then be refined by this method. But maybe it makes more sense to think about that if it happens. What do you think?

joeyballentine replied:
Usually the way we handle this is we have the node named after the action and we have tabs or a dropdown for the implementation and it's options. There's no need to have multiple nodes for all the different ways to align

pifroggi replied:
Okay! Should we combine the average color fix and the wavelet color fix into one node as well?

joeyballentine commented on backend/src/packages/chaiNNer_pytorch/pytorch/processing/rife/IFNet_HDv3_v4_14_align.py:
```diff
@@ -0,0 +1,187 @@
+#logging.basicConfig(filename='debug.log', level=logging.DEBUG, format='%(asctime)s %(levelname)s:%(message)s')
+
+
+#device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
can you get rid of all these random comments?

pifroggi replied:
Okay, should we remove the links/credits to the original code as well?

joeyballentine commented:
Please run the ruff formatter on the code as well. Thanks

joeyballentine commented on backend/src/packages/chaiNNer_pytorch/pytorch/processing/image_align_rife.py:
```diff
@@ -0,0 +1,211 @@
+        NumberInput(
+            "Alignment Passes",
+            controls_step=1,
+            maximum=1000,
```
This needs to have a minimum of 1. 0 causes an error.

joeyballentine commented on backend/src/packages/chaiNNer_pytorch/pytorch/processing/image_align_rife.py:
```diff
@@ -0,0 +1,211 @@
+def download_model(download_url, model_path, model_file, zip_inner_path):
+    model_dir = Path(model_path)
+    model_dir.mkdir(parents=True, exist_ok=True)
+    zip_path = model_dir / "model.zip"
```
Instead of this whole sequence you do here, I would prefer if you did something like this: https://github.com/chaiNNer-org/chaiNNer/blob/bfaac0a516057d0898154618b3b28b6109a4eafa/backend/src/packages/chaiNNer_pytorch/pytorch/restoration/upscale_face.py#L169

This gets the appdata/roaming path where we store everything else. The code currently just puts it near wherever chaiNNer is running.

pifroggi commented:
I'll look into the ruff formatter you mentioned and do the other changes likely later today.

Splendide-Imaginarius commented:
> The Wavelet Color Fix functions are from [sd-webui-stablesr](https://github.com/pkuliyi2015/sd-webui-stablesr/blob/master/srmodule/colorfix.py).

Does this PR also include the `adain_color_fix` function? If not, is there any reason it's excluded?

Splendide-Imaginarius commented:
How does the Wavelet Color Fix handle fully transparent pixels? E.g. if you take an image that has some fully transparent regions, and run `magick in.png -background lime -alpha background out.png` on it, and upscale the result, what does Wavelet Color Fix look like? (See https://github.com/chaiNNer-org/chaiNNer/pull/1907 for more discussion.)

pifroggi commented:
Hey, that sounds great about your homography alignment! :) The wavelet color fix node currently only takes 3 channel inputs, so if your images have transparency, you would have to use for example a "Split Transparency" node and then recombine it with the output again.

I've not included the Adain Color Fix as it didn't produce good results on my use case, but I didn't do a lot of testing on it outside of that, sorry about that. It works quite differently, more similarly to the "Color Transfer" node in chainner. As in the content of the two images can be completely different and it tries to match the overall color palette.
While the Wavelet Color Fix is very similar to the Average Color Fix, as in the content of the two images need to be in the same place spatially.

@joeyballentine Since the two alignment nodes will stay separate for now, I'm guessing the Average Color Fix and Wavelet Color Fix (which requires pytorch) will as well. Regarding the "Image Align with Rife" name, let me know if you would still prefer to use a different naming scheme.

We could maybe either describe it a bit more broadly with something like:
Fast Image Alignment
Refine Image Alignment

Or keep it more specific to which methods are used with:
Homography Image Alignment
Rife Image Alignment

joeyballentine commented:
I would prefer to keep the implementation detail out of the node name

pifroggi commented:
I agree, sometimes it's just hard to find good words to differentiate them when there is more than one and I was unsure if "fast" and "refine" are good fits here. For now I've left it with "Align Image to Reference" and when @Splendide-Imaginarius's node comes it, we can change that if needed.

Other changes:
removed all the commented out unneeded code
added minimum of 1 where needed
download now goes into appdata/roaming/chaiNNer/python/rife_v4.14/weights
ran ruff

Splendide-Imaginarius commented:
Is this using a standard RIFE model, or has the model been modified in some way? Is there a reason why you're not just letting the user provide the model file, so that the user can use different RIFE models per their preference? I'd be a lot more comfortable with having a node that can load a RIFE model from an arbitrary file (just like loading upscale models), so that the user can pick their favorite model (or train their own); the behavior here feels way too magic and inflexible for my tastes.

Splendide-Imaginarius commented:
> The wavelet color fix node currently only takes 3 channel inputs, so if your images have transparency, you would have to use for example a "Split Transparency" node and then recombine it with the output again.

Yeah I know, I was mainly asking whether the end result looks OK when the alpha channel has been split out and recombined. Average Color Fix needed a specific workaround (masking out the transparent pixels when calculating averages) to avoid contaminating the image with the RGB data from the transparent pixels. It sounds like Wavelet Color Fix may be less susceptible to that issue, but I was just curious if you had tried it. It's not a big deal if you haven't tried it, I can test it myself later, and it's not a reason to block the PR.

pifroggi commented:
Hey, I just wanted to check if something would still be missing from my side. I saw there is still some issue with ruff (or is this from the older commits?), but when I do "ruff check --fix filename" I'm getting "All checks passed!".

Sorry if I'm misinterpreting the Github UI.

joeyballentine commented:
I'll just fix that later when I'm and to and then this should be good I think

joeyballentine commented:
Thanks again for this and sorry it took me so long to do the remaining fixes and review.

pifroggi commented:
Awesome and no worries, I'm glad it worked out! Looking forward to the next release! :)
